File: java/src/main/java/problems101_200/WordBreak2.java
package problems101_200;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

enum WordBreak2 {
    DEFAULT {
        @Override
        List<String> wordBreak(String s, List<String> wordDict) {
            return recursiveWordBreak(s, wordDict, new HashMap<>());
        }

        List<String> recursiveWordBreak(
            String s,
            List<String> wordDict,
            Map<String, List<String>> map
        ) {
            if (map.containsKey(s)) {
                return map.get(s);
            }
            List<String> result = new ArrayList<>();
            for (String word : wordDict) {
                if (!s.startsWith(word)) {
                    continue;
                }
                String next = s.substring(word.length());
                if (next.isEmpty()) {
                    result.add(word);
                    continue;
                }
                for (String sub : recursiveWordBreak(next, wordDict, map)) {
                    result.add(word + " " + sub);
                }
            }
            map.put(s, result);
            return result;
        }
    };

    abstract List<String> wordBreak(String s, List<String> wordDict);
}

File: kotlin/src/main/kotlin/problems101_200/WordBreak2.kt
package problems101_200

enum class WordBreak2 {
    DEFAULT {
        override fun wordBreak(s: String, wordDict: List<String>): List<String> =
            recursiveWordBreak(s, wordDict, hashMapOf())

        fun recursiveWordBreak(
            s: String,
            wordDict: List<String>,
            map: MutableMap<String, List<String>>,
        ): List<String> {
            if (map.containsKey(s)) {
                return map[s]!!
            }
            val result = arrayListOf<String>()
            for (word in wordDict) {
                if (!s.startsWith(word)) {
                    continue
                }
                val next = s.substring(word.length)
                if (next.isEmpty()) {
                    result += word
                    continue
                }
                for (sub in recursiveWordBreak(next, wordDict, map)) {
                    result += "$word $sub"
                }
            }
            map[s] = result
            return result
        }
    },
    ;

    abstract fun wordBreak(s: String, wordDict: List<String>): List<String>
}
